### PR TITLE
Fix Patch Provider pulling in parallel

### PR DESCRIPTION
**Pull request by LuAbelt**
Adds some safeguards to avoid pulling concurrently when running experiments on the cluster

Fixes https://github.com/se-sic/VaRA/issues/1038

### CODE REVIEW

File: varats-core/varats/utils/filesystem_util.py [hunk 1]
```diff
@@ -13,3 +15,18 @@ def __init__(self, folder: tp.Union[Path, str]) -> None:
...
             "but was already present."
         )
+
+
+@contextmanager
+def lock_file(lock_path: Path, lock_mode: int = fcntl.LOCK_EX) -> tp.Generator:
+    # Ensure that the lock exists
+    lock_path.touch(exist_ok=True)
+
+    open_mode = os.O_RDWR | os.O_CREAT | os.O_TRUNC
```
Comment (boehmseb) [inline varats-core/varats/utils/filesystem_util.py]: Shouldn't `os.O_CREAT` already ensure that the lock file exists?

File: varats-core/varats/utils/filesystem_util.py [hunk 2]
```diff
@@ -13,3 +15,18 @@ def __init__(self, folder: tp.Union[Path, str]) -> None:
             f"Folder: '{str(folder)}' should be created "
             "but was already present."
         )
+
+
+@contextmanager
+def lock_file(lock_path: Path, lock_mode: int = fcntl.LOCK_EX) -> tp.Generator:
```
Comment (boehmseb) [inline varats-core/varats/utils/filesystem_util.py:21]: Interesting that mypy does not complain about the missing generic parameters to `tp.Generator` :thinking:

File: varats-core/varats/utils/filesystem_util.py (hunk 1 above)
Comment (LuAbelt) [reply varats-core/varats/utils/filesystem_util.py]: True, I will remove the `touch`

File: varats-core/varats/provider/patch/patch_provider.py [hunk 3]
```diff
@@ -237,12 +238,10 @@ class PatchProvider(Provider):
...
 
-        # BB only performs a fetch so our repo might be out of date
-        pull_current_branch(self._get_patches_repository_path())
+        self._update_local_patches_repo()
+        repo_path = self._get_patches_repository_path()
 
-        patches_project_dir = Path(
-            self._get_patches_repository_path() / self.project.NAME
-        )
+        patches_project_dir = Path(repo_path / self.project.NAME)
```
Comment (vulder) [inline varats-core/varats/provider/patch/patch_provider.py]:
```suggestion
        patches_project_dir = repo_path / self.project.NAME
```
repo_path is already a path, so if NAME is a string you can directly add it to the path and get a path

**Review by vulder — APPROVED** [review]
Otherwise, following the docs, this looks fine.

**Review by vulder — APPROVED** [review]

**Review by boehmseb — APPROVED** [review]

**Pull request merged**